File: long_lines.js
var is_debug = true;
var input_stdin = "";
var input_cursor = 0;

function main() {
    let number_of_people = nextInt();
    let h_of_0 = nextInt();
    let var_a = nextInt();
    let var_c = nextInt();
    let var_q = nextInt();

    let heights = [h_of_0];

    let counter = 1

    while (counter < number_of_people) {
        let height = (((var_a * heights[counter - 1]) + var_c) % var_q)
        heights.push(height);
        counter += 1
    }

    if (is_debug) {
        heights = [2012, 2431, 2496, 2328, 1258];
    }

    if (is_debug) {
        //heights = [400, 200, 200, 300, 100, 600, 150, 50, 130, 900, 30];
    }

    heights.reverse();

    if (is_debug) {
        console.log(heights)
    }

    let noticable = 0;

    function taller_between(start, end, height) {
        let temp = heights.filter((_, ind) => ind >= start && ind <= end);
        for (let i = 0; i <= temp.length; i++) {
            if (temp[i] > height) return temp[i];
        }
        return -1;
    }

    function all_shorter(start, end, height) {
        let temp = heights.filter((_, ind) => ind >= start && ind <= end);
        for (let i = 0; i <= temp.length; i++) {
            if (temp[i] > height) return false;
        }
        return true;
    }

    function all_remaining_shorter(k, height) {
        let temp = heights.filter((_, ind) => ind >= k);
        for (let i = 0; i <= temp.length; i++) {
            if (temp[i] > height) return false;
        }
        return true;
    }

    for (let i = 0; i < heights.length; i++) {
        let current_person_height = heights[i];

        for (let k = i + 1; k < heights.length; k++) {
            let target_person_height = heights[k];

            if (is_debug) {
                console.log('range:', heights.filter((_, ind) => ind >= i && ind <= k));
            }

            if (k - i < 2) {
                noticable++;
            } else {
                //let tallest = tallest_in_range(i, k);

                if (current_person_height >= target_person_height) {
                    if (all_shorter(i, k, target_person_height)) {
                        if (is_debug) {
                            console.log('noticable')
                        }
                        noticable++;
                    }
                }
                // else if (current_person_height < target_person_height && k - i === 0) {
                //     noticable++;
                // } 
                else if (current_person_height < target_person_height) {

                    let taller = taller_between(i, k, target_person_height);

                    if (taller === -1) {
                        if (is_debug) {
                            console.log('noticable')
                        }
                        noticable++;
                    } else {
                        if (is_debug) {
                            console.log(taller, 'is between', heights[i], heights[k]);
                        }
                    }

                    if (all_remaining_shorter(k, target_person_height)) {
                        break;
                    }
                }
            }


        }
        if (is_debug) {
            console.log('------------------')
        }
    }

    console.log(noticable)
}

if (is_debug) {
    main();
} else {
    process.stdin.resume();
    process.stdin.setEncoding('ascii');
    process.stdin.on('data', function (data) {
        input_stdin += data;
    });
    process.stdin.on('end', function () { main(); });
}

// default parsers for JS.
function nextInt() {
    return parseInt(nextString());
}

function nextFloat() {
    return parseFloat(nextString());
}

function nextString() {
    var next_string = "";
    clearWhitespaces();
    while (input_cursor < input_stdin.length && !isWhitespace(input_stdin[input_cursor])) {
        next_string += input_stdin[input_cursor];
        input_cursor += 1;
    }
    return next_string;
}

function nextChar() {
    clearWhitespaces();
    if (input_cursor < input_stdin.length) {
        return input_stdin[input_cursor++];
    } else {
        return '\0';
    }
}

function isWhitespace(character) {
    return ' \t\n\r\v'.indexOf(character) > -1;
}

function clearWhitespaces() {
    while (input_cursor < input_stdin.length && isWhitespace(input_stdin[input_cursor])) {
        // ignore the next whitespace character
        input_cursor += 1;
    }
}

File: long_lines.py
# number_of_people h_of_0 a c q
is_debug = True

user_input = None

if is_debug:
    user_input = '5 2012 37 31 2573'.split(' ')
else:
    user_input = input().split(' ')


number_of_people = int(user_input[0])
h_of_0 = int(user_input[1])
var_a = int(user_input[2])
var_c = int(user_input[3])
var_q = int(user_input[4])

heights = [h_of_0]

counter = 1
while counter < number_of_people:
    height = (((var_a * heights[counter - 1]) + var_c) % var_q)
    heights.append(height)
    counter += 1

if is_debug:
    heights = [400, 200, 200, 300, 100, 600, 150, 50, 130, 900, 30]

heights.reverse()

if is_debug:
    print(heights)

noticable_counter = 0


def has_taller_inbetween(target_person, start, end):
    if is_debug:
        print(heights[start:end+1])

    sub_array = heights[start:end+1]
    for person in sub_array:
        if person > target_person:
            return person
    return -1


def compare_sub_array(person, start, end):
    global noticable_counter
    local_tallest = person
    local_noticable = 0
    for i, target_person in enumerate(heights[start:end]):
        if target_person < person:
            if i == 0:
                local_noticable += 1
            else:
                taller_person = has_taller_inbetween(
                    target_person, start, start+i)
                if taller_person == -1:
                    local_noticable += 1
                else:
                    if is_debug:
                        print('between', person, 'and', target_person,
                              'taller person is', taller_person)
        elif target_person > person and target_person > local_tallest:
            local_noticable += 1
            local_tallest = target_person
        else:
            break

    # if is_debug:
    #     print(local_noticable)

    noticable_counter += local_noticable


for index, person in enumerate(heights):
    compare_sub_array(person, index+1, len(heights))


print(noticable_counter)

# for start_index_b, _ in enumerate(heights):
#     if start_index_b > start_index:
#         tallest = 0
#         print('round', start_index, range(start_index_b, len(heights)))
#         for index_b in range(start_index_b, len(heights)):
#             current_target = heights[index_b]
#             # print(current_target)
#             if current_target > current_person and current_target > tallest:
#                 noticable_counter += 1
#                 tallest = current_target
